cmd/registry-migrator/main.go

```diff
@@ -0,0 +1,212 @@
+
+	flag.Parse()
+
+	clientv3.SetLogger(grpclog.NewLoggerV2(os.Stderr, os.Stderr, os.Stderr))
+
+	if oldRegistryPrefix == "" {
+		log.Fatalln("registry-prefix cannot be empty")
+	}
+
+	if newRegistryPrefix == "" {
+		log.Fatalln("new-refistry-prefix cannot be empty")
+	}
+
+	if backup == "" {
+		log.Fatalln("backup-file cannot be empty")
+	}
+
+	eps := strings.Split(endpoints, ",")
+	if len(eps) == 0 {
+		log.Fatalln("at least one endpoint must be provided")
+	}
+
+	cc := clientv3.Config{
+		Endpoints:   eps,
+		DialTimeout: dialTimeout,
+	}
+
+	switch {
+	case caFile != "" && certFile != "" && keyFile != "":
+		tlsInfo := transport.TLSInfo{
+			TrustedCAFile: caFile,
+			CertFile:      certFile,
+			KeyFile:       keyFile,
+		}
+
+		tls, err := tlsInfo.ClientConfig()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		cc.TLS = tls
+	case caFile == "" && certFile == "" && keyFile == "":
+	default:
+		log.Fatalf("--ca, --cert and --key must all be specified when one is provided")
+	}
+
+	cli, err := clientv3.New(cc)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer cli.Close()
+
+	lg, err := zap.NewProduction()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	dummy, err := cli.Get(context.TODO(), "/dummy-foo-bar", clientv3.WithKeysOnly())
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Printf("compacting to revision %d", dummy.Header.Revision)
+
+	_, err = cli.Compact(context.TODO(), dummy.Header.Revision, clientv3.WithCompactPhysical())
+	if err != nil {
+		if err == rpctypes.ErrCompacted {
+			log.Println("etcd is already compacted to the latest revision")
+		} else {
+			log.Fatal(err)
+		}
+	}
+
+	log.Printf("compaction successful")
+
+	sp := snapshot.NewV3(lg)
+
+	err = sp.Save(context.TODO(), cc, backup)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	old, err := cli.Get(context.TODO(), oldRegistryPrefix,
+		clientv3.WithPrefix(),
+		clientv3.WithKeysOnly(),
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Printf("Starting from revision %d", old.Header.Revision)
+
+	for _, k := range old.Kvs {
+		key := string(k.Key)
+		newKey := newRegistryPrefix + strings.TrimPrefix(key, oldRegistryPrefix)
+		log.Printf("migrating %q => %q", key, newKey)
+
+		out, err := cli.Get(context.TODO(), key)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		if len(out.Kvs) != 1 {
+			log.Fatalf("key %q is missing", key)
+		}
+
+		compares := []clientv3.Cmp{clientv3.Compare(clientv3.ModRevision(key), "=", out.Kvs[0].ModRevision)}
+		if !force {
+			compares = append(compares, clientv3util.KeyMissing(newKey))
+		}
+
+		txn, err := cli.Txn(context.TODO()).
+			If(compares...).
+			Then(clientv3.OpPut(newKey, string(out.Kvs[0].Value))).
+			Commit()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		if !txn.Succeeded {
+			if !force {
+				log.Fatalf(keyExistsErrorMsg, *txn.OpResponse().Txn())
+			}
+		}
+	}
+
+	if delete {
+		log.Println("Deleting resources. You have 5 seconds to cancel")
+		time.Sleep(time.Second * 5)
+
+		// don't explicitly delete the entire key prefix
```

Comment:
Will you elaborate why didn't you used delete api with prefix options here?

Reply:
Because you might accidentally delete keys which already exist at this prefix.

Reply:
Implement algo also does the same.

e.g. if have before migration
```
foo val
foo1 val1
```
And after migration with following command

```
go run main.go --backup-file ./etcd.backup --old-registry-prefix="foo" --new-registry-prefix "bar"  --delete
```
Both implemented and alternative with `del api` will also have same output. 
```
bar val
bar1 val1
```
Anyways i'm fine with both. As long as both are working fine. But i prefer using existing single api call over implementing it again by yourself.